src/garage/torch/algos/sac.py

```diff
@@ -0,0 +1,304 @@
+        if self.use_automatic_entropy_tuning:
+            alpha_loss = (-(self.log_alpha) * (log_pi.detach() + self.target_entropy)).mean()
+        return alpha_loss
+
+    def actor_objective(self, obs, log_pi, new_actions):
+        alpha = self.log_alpha.detach().exp()
+        min_q_new_actions = torch.min(self.qf1(torch.Tensor(obs), torch.Tensor(new_actions)), 
+                            self.qf2(torch.Tensor(obs), torch.Tensor(new_actions)))
+        policy_objective = ((alpha * log_pi) - min_q_new_actions.flatten()).mean()
+        return policy_objective
+
+    def critic_objective(self,samples):
+        '''
+        QF Loss
+        '''
+        obs = samples["observation"]
+        actions = samples["action"]
+        rewards = samples["reward"]
+        terminals = samples["terminal"]
+        next_obs = samples["next_observation"]
+
+        alpha = self.log_alpha.detach().exp()
+
+        q1_pred = self.qf1(torch.Tensor(obs), torch.Tensor(actions))
+        q2_pred = self.qf2(torch.Tensor(obs), torch.Tensor(actions))
+
+        new_next_actions_dist = self.policy(torch.Tensor(next_obs))
+        new_next_actions_pre_tanh, new_next_actions = new_next_actions_dist.rsample(return_pre_tanh_value=True)
+        new_log_pi = new_next_actions_dist.log_prob(value=new_next_actions, pre_tanh_value=new_next_actions_pre_tanh)
+
+        target_q_values = torch.min(
+            self.target_qf1(torch.Tensor(next_obs), new_next_actions),
+            self.target_qf2(torch.Tensor(next_obs), new_next_actions)
+        ).flatten() - (alpha * new_log_pi)
+        with torch.no_grad():
+            q_target = self.reward_scale * torch.Tensor(rewards) + (1. - torch.Tensor(terminals)) * self.discount * target_q_values
+        qf1_loss = F.mse_loss(q1_pred.flatten(), q_target)
+        qf2_loss = F.mse_loss(q2_pred.flatten(), q_target)
+
+        return qf1_loss, qf2_loss
+
+    def update_targets(self):
+        """Update parameters in the target q-functions."""
+        # update for target_qf1
+        target_qfs = [self.target_qf1, self.target_qf2]
+        qfs = [self.qf1, self.qf2]
+        for target_qf, qf in zip(target_qfs, qfs):
+                for t_param, param in zip(target_qf.parameters(),
+                                            qf.parameters()):
+                        t_param.data.copy_(t_param.data * (1.0 - self.tau) +
+                                        param.data * self.tau)
+
+    def optimize_policy(self, itr, gradient_step, samples):
+        """ Optimize the policy based on the policy objective from the sac paper.
+
+        Args:
+            itr (int) - current training iteration
+            samples() - samples recovered from the replay buffer
+        Returns:
+            None
+        """
+
+        obs = samples["observation"]
+
+        # ================update q functions===============#
+        qf1_loss, qf2_loss = self.critic_objective(samples)
+        if(self.gradient_steps == 1):
```

Comment:
Can you refactor out this `if ... elif` logic into a helper method (instead of writing all of the logging code twice)?